server/src/main/resources/UI_Interaction.puml

```diff
@@ -0,0 +1,54 @@
+@startuml
+participant TerraUI
+box "Leo VM"
+    participant JupyterExtension
+    participant JupyterAutoSaveExtension
+    participant Welder
+end box
+participant GCS
+
+TerraUI -> GCS ++: isLocked?
+alt No Lock or locked by current user
+    GCS -> TerraUI:
+    note left: Enable EDIT button
+else LockBy someone else
+    return
+    note left: Disable EDIT mode button
+end
+TerraUI -> Welder: POST /storageLinks
+note left: update storagelink for a given notebook(both SAFE mode and EDIT mode)
+TerraUI -> Welder: localize
+note left: localize file to corresponding location
+
+JupyterExtension -> Welder ++ #F8C471: /checkMeta (syncMode)
+alt 200 SyncMode.EDIT
+    Welder -> JupyterExtension: EDIT mode
+    JupyterExtension -> Welder: /acquireLock
+
+    alt 200 successfully acquire Lock
+        Welder -> JupyterExtension: No Lock or locked by current user
+        loop every 30 seconds?
+            JupyterExtension -> Welder ++ #F4D03F: /checkMeta
+            alt SyncStatus.LIVE
+                Welder -> JupyterExtension: file is live
+            else SyncStatus.REMOTE_NOT_FOUND
+                Welder -> JupyterExtension: remote file deleted
+                note left: Do we assume user will just want to /safeDelocalize?
```

Comment:
The UI extension could offer this as an option. We shouldn't assume this though and automatically save - we would want to let the user know they are desycnhronized.

Reply:
If `SyncStatus.REMOTE_NOT_FOUND` happens in `/safeDelocalize`, shall we do nothing then?

Reply:
We should delocalize. Can set a generation precondition which matches this (I think 0 for creating new files).

Reply:
What's the point of precondition in the case of `RMOTE_NOT_FOUND`?

Reply:
To avoid two users writing to the same location simultaneously. Very unlikely happen, but doesn't hurt to set it.